Remaining STJ work for .NET 8 (#37806)
dotnet/docs: docs/standard/serialization/system-text-json/source-generation-modes.md

The optimized code doesn't support all of the serialization features that `JsonSerializer` supports. The serializer detects whether the optimized code can be used and falls back to default serialization code if unsupported options are specified. For example, <xref:System.Text.Json.Serialization.JsonNumberHandling.AllowReadingFromString?displayProperty=nameWithType> isn't applicable to writing, so specifying this option doesn't cause a fallback to default code.

The following table shows which options in `JsonSerializerOptions` are supported by the optimized serialization code:

| Serialization option                                                   | Supported by optimized code |
|------------------------------------------------------------------------|-----------------------------|

> | <xref:System.Text.Json.JsonSerializerOptions.AllowTrailingCommas>      | ❌                          |

Reviewer: ```suggestion
| <xref:System.Text.Json.JsonSerializerOptions.AllowTrailingCommas>      | ✔️                          |
```

| <xref:System.Text.Json.JsonSerializerOptions.Converters>               | ❌                          |

> | <xref:System.Text.Json.JsonSerializerOptions.DefaultBufferSize>        | ❌                          |

Reviewer: ```suggestion
| <xref:System.Text.Json.JsonSerializerOptions.DefaultBufferSize>        | ✔️                          |
```

| <xref:System.Text.Json.JsonSerializerOptions.DefaultIgnoreCondition>   | ✔️                         |
| <xref:System.Text.Json.JsonSerializerOptions.DictionaryKeyPolicy>      | ❌                          |
| <xref:System.Text.Json.JsonSerializerOptions.Encoder>                  | ❌                          |
| <xref:System.Text.Json.JsonSerializerOptions.IgnoreNullValues>         | ❌                          |
| <xref:System.Text.Json.JsonSerializerOptions.IgnoreReadOnlyFields>     | ✔️                         |
| <xref:System.Text.Json.JsonSerializerOptions.IgnoreReadOnlyProperties> | ✔️                         |
| <xref:System.Text.Json.JsonSerializerOptions.IncludeFields>            | ✔️                         |

> | <xref:System.Text.Json.JsonSerializerOptions.MaxDepth>                 | ❌                          |

Reviewer: ```suggestion
| <xref:System.Text.Json.JsonSerializerOptions.MaxDepth>                 | ✔️                          |
```

| <xref:System.Text.Json.JsonSerializerOptions.NumberHandling>           | ❌                          |

> | <xref:System.Text.Json.JsonSerializerOptions.PropertyNameCaseInsensitive> | ❌                       |

Reviewer: This is a deserialization-only setting. I'm not sure if it makes sense to include when discussing fast-path serialization, but perhaps it's helpful in emphasizing that there's not fast-path deserialization.

| <xref:System.Text.Json.JsonSerializerOptions.PropertyNamingPolicy>     | ✔️                         |

> | <xref:System.Text.Json.JsonSerializerOptions.ReadCommentHandling>      | ❌                          |

Reviewer: Likewise, this is deserialization-only.

| <xref:System.Text.Json.JsonSerializerOptions.ReferenceHandler>         | ❌                          |
| <xref:System.Text.Json.JsonSerializerOptions.TypeInfoResolver>         | ✔️                         |

> | <xref:System.Text.Json.JsonSerializerOptions.UnknownTypeHandling>      | ❌                          |

Reviewer: Likewise, deserialization-only.

| <xref:System.Text.Json.Serialization.JsonNumberHandlingAttribute> | ❌                         |
| <xref:System.Text.Json.Serialization.JsonPolymorphicAttribute>    | ✔️                         |
| <xref:System.Text.Json.Serialization.JsonPropertyNameAttribute>   | ✔️                         |
| <xref:System.Text.Json.Serialization.JsonPropertyOrderAttribute>  | ❌                         |
| <xref:System.Text.Json.Serialization.JsonRequiredAttribute>       | ✔️                         |

> If a non-supported option or attribute is specified for a type, the serializer falls back to the default `JsonSerializer` code. In that case, the optimized code isn't used when serializing that type but may be used for other types. Therefore it's important to do performance testing with your options and workloads to determine how much benefit you can actually get from serialization-optimization mode. Also, the ability to fall back to `JsonSerializer` code requires metadata-collection mode. If you select only serialization-optimization mode, serialization might fail for types or options that need to fall back to `JsonSerializer` code.

Reviewer: ```suggestion
If a non-supported option or attribute is specified for a type, the serializer falls back to metadata mode, assuming that metadata initialization code has been generated. In that case, the optimized code isn't used when serializing that type but may be used for other types. Therefore it's important to do performance testing with your options and workloads to determine how much benefit you can actually get from serialization-optimization mode. Also, the ability to fall back to `JsonSerializer` code requires metadata-collection mode. If you select only serialization-optimization mode, serialization might fail for types or options that need to fall back to `JsonSerializer` code.
```
